prow/pod-utils/gcs/upload.go

```diff
@@ -171,12 +197,51 @@ func (w gcsObjectWriter) ApplyAttributes(attrs *storage.ObjectAttrs) {
 	w.Writer.ObjectAttrs = *attrs
 }
 
+type openerObjectWriter struct {
+	pkgio.Opener
+	Bucket      string
+	Dest        string
+	writeCloser pkgio.WriteCloser
```

Comment:
if we ever want to use this for GCS as well or pass any kind of opts, we need to change the `Writer` method in the `io` package to accept some kind of opts. Could probably be a simplistic struct that only contains GCS options and gets extended in the future if the need arises. We can do this backwards-compatible by leveraging the functional opts pattern.

Reply:
That was the reason why I added the blob.WriterOptions in some variant of the previous. Although I admit that the functional opts pattern is better for that kind of requirement.

Anyway. I think the main problem is that we cannot fullfill both use cases at the same time:

* The intent of this package is to:
   *replace https://golang.org/pkg/io/ imports with k8s.io/test-infra/pkg/io and have the code work the same as before (i.e. not changing the opener interface)

* Make the writer customizable so we can set WriterOptions or Attributes when writing

Reply:
> Anyway. I think the main problem is that we cannot fullfill both use cases at the same time:

I think we can, because the opts are set when constructing the `io.WriteCloser` in the `Writer` method, not when calling `Write`, only the latter is part of the `io.Writer`/`io.WriteCloser` interface.

I _think_ @fejta wasn't generally opposed to allowing opts, just to using `*blob.WriterOptions` for it...?

Reply:
> I think we can, because the opts are set when constructing the io.WriteCloser in the Writer method, not when calling Write, only the latter is part of the io.Writer/io.WriteCloser interface.

This would mean during the construction of the opener right? If that's the case the problem is that we don't know which files we want to write when the opener is constructed and the properties are different per file.

I would prefer adding options for Write(), if that's an option

Reply:
In theory this sounds good, in practise its not possible to change opts on a `Writer`: https://godoc.org/gocloud.dev/blob#Writer

I am actually surprised the GCS implementation allows this, because that means its not threadsafe. For us, setting the `Opts` when constructing the `Writer` is completely fine as we construct a `Writer` per file anyways (which makes sense due to thread safety and to have a per-file timeout)

Reply:
Yup. The only question is if we want to change the signature of the Writer() method to include this options in any way. The only alternative would be to create an opener every time when we need an writer and set the options in Opener. But I think that would be really bad and breaks a lot of stuff :)

Reply:
Yes, we do, this doesn't belong into the `Opener`

Reply:
*push* fejta (if possible :))

Reply:
/assign fejta
WDYT?

Reply:
*push* @fejta

Reply:
In addition to size, it looks like these are the field we set:

https://github.com/kubernetes/test-infra/blob/d2e2e04d1b0335335ae006973ba606e95f585af9/prow/pod-utils/gcs/metadata.go#L54-L76

https://github.com/kubernetes/test-infra/blob/d2e2e04d1b0335335ae006973ba606e95f585af9/prow/pod-utils/gcs/upload.go#L129

It does seem like adding options like this to the opener.Writer interface is the most logical place.

I like the []Option pattern, which should allow us to distinguish between parameters that are common between providers (size, content-type) from possibly provider specific functionality (if-generation-match, chunksize)

Reply:
I implemented the WriterOptions

Some notes:
* gocloud uses BufferSize as ChunkSize when the gcs implemention of gocloud is used
* I only implemented ContentEncoding, ContentType and MetaData as these are the only attributes we're currently using. The disadvantage is that if we set other attributes e.g. `FileUploadWithAttributes` they are not used. I didn't want to expose `storage.ObjectAttrs` in the opener package. We could:
  * keep it as is
  * add a WithGcsObjectAttrs option (won't solve the problem for other providers)
  * try to implement all options (which are a lot)
  * refactor `FileUploadWithAttributes` to accept only `contentType`, `contentEncoding` and `metadata` 
  The last one is my favorite because it has the least potential for errors and isn't to verbose. In this case, we should probably also refactor `gcs.AttributesFromFileName` because it effectively only calculates `contentType` and `contentEncoding`.
*

Reply:
:+1: for only allowing to set the options we actually use

Reply:
I migrated from storage.ObjectAttrs to io.WriterOptions. Looks cleaner imho